pallets/unified-accounts/src/lib.rs

```diff
@@ -0,0 +1,307 @@
+        /// Ensure no prior mapping exists for the account
+        #[pallet::call_index(1)]
+        pub fn claim_default_evm_address(origin: OriginFor<T>) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            // make sure no prior mapping exists
+            Self::enure_no_mapping(&who, &None)?;
+            // claim default evm address
+            let _ = Self::do_claim_default_evm_address(who)?;
+            Ok(())
+        }
+    }
+}
+
+impl<T: Config> Pallet<T> {
+    /// Ensure no mappings exists for given pair of account/evm_address
+    fn enure_no_mapping(
+        account_id: &T::AccountId,
+        evm_address: &Option<EvmAddress>,
+    ) -> DispatchResult {
+        // ensure account_id and evm address has not been mapped
+        ensure!(
+            !EvmToNative::<T>::contains_key(&account_id),
+            Error::<T>::AlreadyMapped
+        );
+        // This is not required since checking one mapping is sufficent
+        // but this is just for sanity check
+        if let Some(addr) = evm_address {
+            ensure!(
+                !NativeToEvm::<T>::contains_key(addr),
+                Error::<T>::AlreadyMapped
+            );
+        }
+        Ok(())
+    }
+
+    /// Add the given pair to create double mappings
+    fn add_mappings(account_id: T::AccountId, evm_address: EvmAddress) {
+        NativeToEvm::<T>::insert(&evm_address, &account_id);
+        EvmToNative::<T>::insert(&account_id, &evm_address);
+
+        Self::deposit_event(Event::AccountClaimed {
+            account_id,
+            evm_address,
+        });
+    }
+
+    /// Claim the given evm address by providing claim signature
+    fn do_claim_evm_address(
+        account_id: T::AccountId,
+        evm_address: EvmAddress,
+        signature: SignatureOf<T>,
+    ) -> DispatchResult {
+        // recover evm address from signature
+        let address = T::SignatureHelper::verify_signature(&account_id, &signature)
+            .ok_or(Error::<T>::BadSignature)?;
+        log::trace!(
+            target: "account::do_claim_address",
+            "evm_address: {:#?}, recovered: {:#?}", evm_address, address
+        );
+        ensure!(evm_address == address, Error::<T>::InvalidSignature);
+
+        // Check if the default account id already exists for this evm address
+        let default_account_id = T::DefaultAddressMapping::into_account_id(evm_address.clone());
+        if frame_system::Pallet::<T>::account_exists(&default_account_id) {
+            // Transfer all the free balance from old account id to the newly
+            // since this `default_account_id` will no longer be connected to evm address
+            // and users cannot access it.
+            T::Currency::transfer(
+                &default_account_id,
+                &account_id,
+                T::Currency::reducible_balance(&default_account_id, Expendable, Polite),
+                Expendable,
+            )?;
```

Comment:
`pallet-assets` under this account also need to be transferred.

Reply:
That'll be handled by UI.

Reply:
Why leaving it to UI?

Reply:
Mainly because it add complex logic to runtime that we would have to maintain for the same things that UI can also handle (via simple dispatch precompile call).
Also we don't have a _easy_ way to transfer all of user funds from pallet-assets. (no good way to enumerate over user holdings in pallet-assets, no `TransferAll` from orml)

Reply:
Yes, there is no simple & straightforward way to handle transfer all. But without it, this call implementation is not complete. You can even choose to change the impl so that UI handles the base currency transfer as well, which apparently is not the right way.

Reply:
I'll add dev comments mentioning only native balance is handled by pallet, rest should be handled by UI.